### DISCUSSION TITLE
Phase 2 Key Exchange, DH Group not selected

### DISCUSSION THREAD

**Discussion by ptchuba** [strongswan/strongswan#1203] [Q&A] [answered]
Hello,
I'm setting up an implementation that is supposed to be CNSA compliant. Using Ubuntu 22.04 and latest version of Strongswan with Swanctl. My tunnels come up and I can pass traffic just fine. However, on the very first Child SA, the configured ESP DH group is not selected. When the SA is rekeyed though, it gets the right key exchange algorithm. 

This is what my swanctl.conf looks like

``` 
connections {

   vdi-hub {
      remote_addrs = 2.2.2.2
      local_addrs = 1.1.1.1
      local {
         auth = psk
         id = vdi1.whatever.com
      }
      remote {
         auth = psk
         id = hub.whatever.com
      }
      children {
         hub-sa {
            local_ts  = 10.0.0.2/32
            remote_ts = 10.0.0.0/24,10.1.0.0/24,192.168.5.0/24
            if_id_in = 41
            if_id_out = 41
            updown = /usr/local/lib/ipsec/_updown iptables
            esp_proposals = aes256gcm16-ecp384
            mode = tunnel
            start_action = start
         }
      }
      version = 2
      mobike = no
      encap = yes
      proposals = aes256gcm16-prfsha512-ecp384
   }
}
secrets {
   ike-1 {
      id-1 = hub.whatever.com
      secret = ubuntu123
   }
} 
```
The configuration is the same on the other side, just a mirror image. Increased the log level and this is what i see for the first child SA. 

``` 
Aug  3 21:54:52 ubuntu22-server charon: 01[CFG] received proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ
Aug  3 21:54:52 ubuntu22-server charon: 01[CFG] configured proposals: ESP:AES_GCM_16_256/ECP_384/NO_EXT_SEQ
Aug  3 21:54:52 ubuntu22-server charon: 01[CFG] selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ
Aug  3 21:54:52 ubuntu22-server charon: 01[KNL] got SPI cffba7b6 
``` 

Ipsec statusall output below. Notice that the hub-sa{1} ESP section only shows AES_GCM_16_256 with no group. 
``` 
root@ubuntu22-server:~# ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.5, Linux 5.15.0-43-generic, x86_64):
  uptime: 34 seconds, since Aug 08 16:20:30 2022
  malloc: sbrk 2957312, mmap 0, used 1299248, free 1658064
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 3
  loaded plugins: charon aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark stroke vici updown eap-mschapv2 xauth-generic counters
Listening IP addresses:
  1.1.1.1
  10.0.0.2
Connections:
     vdi-hub:  1.1.1.1...2.2.2.2  IKEv2
     vdi-hub:   local:  [vdi1.whatever.com] uses pre-shared key authentication
     vdi-hub:   remote: [hub.whatever.com] uses pre-shared key authentication
      hub-sa:   child:  10.0.0.2/32 === 10.0.0.0/24 10.1.0.0/24 192.168.5.0/24 TUNNEL
Security Associations (1 up, 0 connecting):
     vdi-hub[1]: ESTABLISHED 22 seconds ago, 1.1.1.1[vdi1.whatever.com]...2.2.2.2[hub.whatever.com]
     vdi-hub[1]: IKEv2 SPIs: d2710b5e0ac73298_i* 163f5f940d033cbe_r, rekeying in 3 hours
     vdi-hub[1]: IKE proposal: AES_GCM_16_256/PRF_HMAC_SHA2_512/ECP_384
      hub-sa{1}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: cfefacd6_i cb03021f_o
      hub-sa{1}:  AES_GCM_16_256, 1092 bytes_i (13 pkts, 6s ago), 1092 bytes_o (13 pkts, 6s ago), rekeying in 57 minutes
      hub-sa{1}:   10.0.0.2/32 === 10.0.0.0/24 10.1.0.0/24 192.168.5.0/24
root@ubuntu22-server:~#  
``` 

Questions:
1. Why isn't the DH group sent in the proposals even though it is explicitly configured?
2.  What DH group is used then? The way I understand the protocols, a group has to be chosen for the entropy of the generated keys.
3. If a DH group is used, where can I see it displayed or what command so I can use that as documentation to meet requirements. This will be ideal so no assumptions are made.

Your help will be greatly appreciated.

**Comment by strongX509** [comment, marked as answer]
The session keys of the first CHILD_SA are always derived from the DH secret of the IKE_SA. Only if CREATE_CHILD_SA is used to establlish multiple CHILD_SAs or during the periodic rekeying of the CHILD_SAs, the DH group specified in the ESP proposal will be used for a fresh key exchange.

If you want the session keys of the first CHILD_SA to be independent  of the DH secret of the IKE_SA ,you can establish a so-called *childless* IKE_SA by setting `childless = force` in the connection definition in `swanctl.conf`,  see the [following example scenario](https://strongswan.org/testing/testresults5/ikev2/net2net-childless/).

**Reply by ptchuba** [reply]
Thanks @strongX509. Do you know if this is based on the RFCs or just a Strongswan implementation? In other words can I expect the same thing from a Cisco or Juniper router/firewall?

Thanks

**Reply by strongX509** [reply]
This is based on the IKEv2 [RFC 7296](https://datatracker.ietf.org/doc/html/rfc7296#section-1.2). Since the first CHILD_SA is negotiated via the IKE_AUTH request/response message pair which doesn't have a KE payload of its own, the session keys for the CHILD_SA keys are derived from the SKEYSEED of the IKE_SA.

**Reply by ptchuba** [reply]
Thank you so much @strongX509. Very much appreciated!
